### PR TITLE
added fuzzing support

### PR DESCRIPTION
**Pull request by fuzzy-boiii23a**
This just adds fuzzing support such that if any changes are made to libgumbo it's easy to fuzz these changes before merging something into the main branch.

### CODE REVIEW

**Comment by fuzzy-boiii23a** [comment]
thoughts @flavorjones ?

**Comment by flavorjones** [comment]
@fuzzy-boiii23a Thank you for opening this! I'm traveling this week but will make time to take a look.

**Comment by stevecheckoway** [comment]
SanityCheckPointers (and as an aside, I've been trying to move away from such terminology) doesn't have any side effects and it doesn't return any values. 1. Have you checked that it doesn't get dead stripped? 2. How will you ensure that it doesn't get dead stripped in the future?

The else will definitely get pruned so if you're hoping that will check if the text pointer is accessible, I don't think that will do it.

There are also pointers for attributes that probably need to be checked as well.

**Comment by fuzzy-boiii23a** [comment]
> SanityCheckPointers (and as an aside, I've been trying to move away from such terminology) doesn't have any side effects and it doesn't return any values. 1. Have you checked that it doesn't get dead stripped? 2. How will you ensure that it doesn't get dead stripped in the future?
> 
> The else will definitely get pruned so if you're hoping that will check if the text pointer is accessible, I don't think that will do it.
> 
> There are also pointers for attributes that probably need to be checked as well.

hi @stevecheckoway , thanks for the heads up! i've made some changes and tested them using -03 to ensure i make it into each if else statement and i do and i also added checking attributes

**Comment by fuzzy-boiii23a** [comment]
Hey @flavorjones hope you're going well and had a great trip, can you think of anything else that I could improve to make it easier to use?

**Comment by flavorjones** [comment]
@fuzzy-boiii23a I've not been able to spend much time on OSS over the past few weeks, I appreciate your patience.

**Comment by fuzzy-boiii23a** [comment]
> @fuzzy-boiii23a I've not been able to spend much time on OSS over the past few weeks, I appreciate your patience.

Absolutely no stress at all :)

File: gumbo-parser/fuzzer/build.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
+export SANITIZER_OPTS=""
+export SANITIZER_LINK=""
+
+if [ -z "${LLVM_CONFIG}" ]
+then
+  echo '$LLVM_CONFIG has not been configured, expecting "export LLVM_CONFIG=/usr/bin/llvm-config-12" assuming clang-12 is installed, however any clang version works'
+  exit
+fi
```
Comment (flavorjones) [inline gumbo-parser/fuzzer/build.sh:8]: My (debian-based) system has a bare `llvm-config` which is a symlink to the active installation (llvm-14). Can we simply use that if it's present?

**Review by flavorjones — CHANGES_REQUESTED** [review]
Thank you again for doing this work!! I really appreciate it, and I'm sorry it's taken me so long to review (I've been traveling a lot).

Can you also add a (very brief) summary of how to run the fuzzer to `CONTRIBUTING.md`? I've only ever used a fuzzer in one other project, and the configuration was much different from this.

- What commands should I run?
- What does output look like if no issues are found?
- What does output look like if an issue is found?

File: gumbo-parser/fuzzer/build.sh (hunk 1 above)
Comment (fuzzy-boiii23a) [reply gumbo-parser/fuzzer/build.sh]: fixed and tested

File: gumbo-parser/fuzzer/build.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+if [ ! -d "build" ]
+then
+  mkdir build
+fi
+
+export CC="$(llvm-config-12 --bindir)/clang"
+export CXX="$(llvm-config-12 --bindir)/clang++"
```
[the remark this answers is not in the document]
Comment (fuzzy-boiii23a) [reply gumbo-parser/fuzzer/build.sh]: fixed, nice catch thanks mate

**Comment by fuzzy-boiii23a** [comment]
> Thank you again for doing this work!! I really appreciate it, and I'm sorry it's taken me so long to review (I've been traveling a lot).
> 
> Can you also add a (very brief) summary of how to run the fuzzer to `CONTRIBUTING.md`? I've only ever used a fuzzer in one other project, and the configuration was much different from this.
> 
> * What commands should I run?
> * What does output look like if no issues are found?
> * What does output look like if an issue is found?

Added those changes https://github.com/fuzzy-boiii23a/nokogiri/blob/fuzzing_support/CONTRIBUTING.md#fuzzing-your-gumbo-parser-changes hope you like them and no need to be sorry thanks for checking out the changes! hope your travel went well :)

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -382,7 +383,7 @@ To modify or add a dependency, a few things needs to be in sync:
 A quick summary of what this looks like for you, the developer:
 
 1. edit the `requirements` in the gemspec
-2. run `bundle exec rake vendor_jars` which updates everything under `lib/nokogiri/jruby`
+2. run `bundle exec rake vendor_jars` which updates everything under `lib/nokogiri/jrubfuzzing-your-changesy`
```
Comment (flavorjones) [inline CONTRIBUTING.md:386]: Looks like an accidental edit?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -26,6 +26,7 @@ If you're looking for guidance on filing a bug report or getting support, please
 - [Bumping Java dependencies](#bumping-java-dependencies)
 - [Rake tasks](#rake-tasks)
 - [Making a release](#making-a-release)
+- [Fuzzing your gumbo-parser changes](#fuzzing-your-changes)
```
Comment (flavorjones) [inline CONTRIBUTING.md]: I don't think this link is correct ... I'd recommend running `rake format:toc` to generate the table of contents.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -408,3 +409,37 @@ A quick checklist:
   - [ ] submit a PR to https://github.com/rubysec/ruby-advisory-db
 - [ ] update nokogiri.org
 - [ ] bump `lib/nokogiri/version/constant.rb` to a prerelease version like `v1.14.0.dev`
+
+## Fuzzing your gumbo-parser changes
```
Comment (flavorjones) [inline CONTRIBUTING.md]: Can you please move this section to just under "How to run the tests"?

File: gumbo-parser/fuzzer/build.sh [hunk 6]
```diff
@@ -1,9 +1,13 @@
 export SANITIZER_OPTS=""
 export SANITIZER_LINK=""
 
+if [ -x "$(command -v llvm-config)" ]; then
```
Comment (flavorjones) [inline gumbo-parser/fuzzer/build.sh:4]: I think this block will overwrite `$LLVM_CONFIG` with `llvm-config` even if I've set the env var. I think it should first check if $LLVM_CONFIG before falling back to `llvm-config`.

**Review by flavorjones — CHANGES_REQUESTED** [review]
Thanks for the quick response. I made a few small requests.

Also, I noticed that the user account associated with the last few commits was different, wanted to ask if that was intentional?

File: gumbo-parser/fuzzer/build.sh (hunk 6 above)
Comment (fuzzy-boiii23a) [reply gumbo-parser/fuzzer/build.sh]: fixed

File: CONTRIBUTING.md (hunk 5 above)
Comment (fuzzy-boiii23a) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 4 above)
Comment (fuzzy-boiii23a) [reply CONTRIBUTING.md]: removed and added under running tests anyway

File: CONTRIBUTING.md (hunk 3 above)
Comment (fuzzy-boiii23a) [reply CONTRIBUTING.md]: fixed that typo have no idea how i managed that haha

**Comment by fuzzy-boiii23a** [comment]
> Thanks for the quick response. I made a few small requests.
> 
> Also, I noticed that the user account associated with the last few commits was different, wanted to ask if that was intentional?

Addressed those changes and the user account was unintentional that's a personal account I use for non fuzzing related projects and forgot to update config :)

**Comment by flavorjones** [comment]
@fuzzy-boiii23a I added a few commits primarily to clean up the build.sh script and allow building the fuzzers in parallel.

At this point, I'm able to run the fuzzers, and I've even found a couple of (very small) memory leaks in gumbo!

I will want to do some follow-up work to assemble a corpus that's useful and make that scriptable. What else do you think needs to be done? Anything? Are you OK if I squash-merge this?

**Review by flavorjones — APPROVED** [review]

**Comment by fuzzy-boiii23a** [comment]
> @fuzzy-boiii23a I added a few commits primarily to clean up the build.sh script and allow building the fuzzers in parallel.
> 
> 
> 
> At this point, I'm able to run the fuzzers, and I've even found a couple of (very small) memory leaks in gumbo! (See #3036)
> 
> 
> 
> I will want to do some follow-up work to assemble a corpus that's useful and make that scriptable. What else do you think needs to be done? Anything? Are you OK if I squash-merge this?

Thanks for that and that's great to hear!, I'm happy to assemble a corpus and a dictionary but for now I think everything is fine to be merged. I will get around to doing that next week.

**Comment by fuzzy-boiii23a** [comment]
> @fuzzy-boiii23a I added a few commits primarily to clean up the build.sh script and allow building the fuzzers in parallel.
> 
> At this point, I'm able to run the fuzzers, and I've even found a couple of (very small) memory leaks in gumbo! (See #3036)
> 
> I will want to do some follow-up work to assemble a corpus that's useful and make that scriptable. What else do you think needs to be done? Anything? Are you OK if I squash-merge this?

Alright so i added a dictionary and corpus and added how to use it in CONTRIBUTING.md which greatly improves coverage so i cannot think of any further improvements to make, looks to to merge to me thanks for the support @flavorjones

**Comment by flavorjones** [comment]
@fuzzy-boiii23a Thank you!

Can you help me understand what's in the `gumbo_corpus.zip`? How was it generated, how would I add to it, etc.?

**Comment by fuzzy-boiii23a** [comment]
> @fuzzy-boiii23a Thank you!
> 
> 
> 
> Can you help me understand what's in the `gumbo_corpus.zip`? How was it generated, how would I add to it, etc.?

No problem, I generated that corpus by running the fuzzer overnight with a the html corpus I grabbed from https://github.com/dvyukov/go-fuzz-corpus along with some scraping of html via the web and after running for 24 hours I performed corpus minimization (https://llvm.org/docs/LibFuzzer.html#id25) which will run each file through the fuzzer and if for example 10 files execute the same amount of code in the target, linger will discard 9 files and keep 1 that exercises the same functionality. You could improve the corpus by collecting heaps of html files and then performing minimization of these inputs using https://llvm.org/docs/LibFuzzer.html#id25

**Comment by fuzzy-boiii23a** [comment]
@flavorjones I had a look for the public corpus that google currently uses for fuzzing nokogiri's implementation of gumbo and it can be downloaded from https://storage.googleapis.com/nokogiri-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/nokogiri_parse_fuzzer/public.zip I'd say that this is the best one to use as the corpus is updated and maintained across their daily run against gumbo which is about a billion executions on average

**Comment by fuzzy-boiii23a** [comment]
Hey @flavorjones hope you're doing well, can we please finalise and merge this one to close it off?

**Comment by flavorjones** [comment]
@fuzzy-boiii23a Thanks for the nudge. The only other thing I need you to educate me on is how to manage the oss-fuzz emails I've been receiving to my personal email.

In https://github.com/sparklemotion/nokogiri/discussions/2992 we talked about using the google group I created, https://groups.google.com/g/nokogiri-oss-fuzz. How can I configure the oss-fuzz settings?

**Pull request merged**